system/lib/wasmfs/wasmfs.cpp

```diff
@@ -6,47 +6,85 @@
 // Current Status: Work in Progress.
 // See https://github.com/emscripten-core/emscripten/issues/15041.
 
+#include "file.h"
+#include "file_table.h"
 #include <emscripten/emscripten.h>
 #include <emscripten/html5.h>
+#include <errno.h>
+#include <mutex>
 #include <stdlib.h>
+#include <utility>
 #include <vector>
 #include <wasi/api.h>
 
 extern "C" {
 
-static std::vector<char> fd_write_stdstream_buffer;
+using namespace wasmfs;
+
+long __syscall_dup2(long oldfd, long newfd) {
+  FileTable::Handle fileTable = FileTable::get();
+
+  auto oldOpenFile = fileTable[oldfd];
+  // If oldfd is not a valid file descriptor, then the call fails,
+  // and newfd is not closed.
+  if (!oldOpenFile) {
+    return -(EBADF);
+  }
+
+  if (newfd < 0) {
+    return -(EBADF);
+  }
+
+  if (oldfd == newfd) {
+    return oldfd;
+  }
+
+  // If the file descriptor newfd was previously open, it will just be overwritten silently.
+  fileTable[newfd] = oldOpenFile;
+  return newfd;
+}
+
+long __syscall_dup(long fd) {
+
+  FileTable::Handle fileTable = FileTable::get();
+
+  // Check that an open file exists corresponding to the given fd.
+  auto currentOpenFile = fileTable[fd];
+  if (!currentOpenFile) {
+    return -(EBADF);
+  }
+
+  // Adds given OpenFileState to FileTable entries. Returns fd (insertion index in entries).
+  // If no free space is found, currentOpenFile will be appended to the back of the entries.
+  for (__wasi_fd_t i = 0;; i++) {
+    if (!fileTable[i]) {
+      // Free open file entry.
+      fileTable[i] = currentOpenFile;
+      return i;
+    }
+  }
+
+  return -(EBADF);
+}
 
 __wasi_errno_t __wasi_fd_write(
   __wasi_fd_t fd, const __wasi_ciovec_t* iovs, size_t iovs_len, __wasi_size_t* nwritten) {
```

Comment:
This indentation looks wrong because the continuation is the same level as the body..  if this is what clang-format is doing I think we might need to tweak the settings).